airsonic-main/src/main/java/org/airsonic/player/service/MediaScannerService.java

```diff
@@ -145,68 +173,76 @@ public synchronized void scanLibrary() {
+                .thenRunAsync(() -> playlistService.importPlaylists(), pool)
+                .thenRunAsync(() -> mediaFileDao.checkpoint(), pool)
+                .thenRun(() -> pool.shutdown())
+                .thenRun(() -> scanning = false);
     }
 
     private void doScanLibrary() {
         LOG.info("Starting to scan media library.");
-        Instant lastScanned = Instant.now().truncatedTo(ChronoUnit.SECONDS);
+        Instant lastScanned = Instant.now();
         LOG.debug("New last scan date is {}", lastScanned);
 
         try {
-
             // Maps from artist name to album count.
-            Map<String, Integer> albumCount = new HashMap<String, Integer>();
+            Map<String, AtomicInteger> albumCount = new ConcurrentHashMap<>();
+            Map<String, Artist> artists = new ConcurrentHashMap<>();
+            Map<MediaFile, Album> albums = new ConcurrentHashMap<>();
             Genres genres = new Genres();
 
-            scanCount = 0;
+            scanCount.set(0);
             statistics.reset();
 
             mediaFileService.setMemoryCacheEnabled(false);
             indexManager.startIndexing();
 
-            mediaFileService.clearMemoryCache();
-
             // Recurse through all files on disk.
-            for (MusicFolder musicFolder : settingsService.getAllMusicFolders()) {
-                MediaFile root = mediaFileService.getMediaFile(musicFolder.getPath(), false);
-                scanFile(root, musicFolder, lastScanned, albumCount, genres, false);
-            }
-
+            ForkJoinTask.invokeAll(
+                    settingsService.getAllMusicFolders()
+                        .parallelStream()
+                        .map(musicFolder -> ForkJoinTask.adapt(() -> scanFile(mediaFileService.getMediaFile(musicFolder.getPath(), false), musicFolder, lastScanned, albumCount, artists, albums, genres, false)))
+                        .collect(Collectors.toList()));
+            
             // Scan podcast folder.
             File podcastFolder = new File(settingsService.getPodcastFolder());
             if (podcastFolder.exists()) {
-                scanFile(mediaFileService.getMediaFile(podcastFolder), new MusicFolder(podcastFolder, null, true, null),
-                         lastScanned, albumCount, genres, true);
+                ForkJoinTask.invokeAll(ForkJoinTask.adapt(() -> scanFile(mediaFileService.getMediaFile(podcastFolder), new MusicFolder(podcastFolder, null, true, null),
+                         lastScanned, albumCount, artists, albums, genres, true)));
             }
 
             LOG.info("Scanned media library with " + scanCount + " entries.");
 
-            LOG.info("Marking non-present files.");
-            mediaFileDao.markNonPresent(lastScanned);
-            LOG.info("Marking non-present artists.");
-            artistDao.markNonPresent(lastScanned);
-            LOG.info("Marking non-present albums.");
-            albumDao.markNonPresent(lastScanned);
```

Comment:
goes to the db serially for each of the tasks

we can just open 3 parallel connections to the db and concurrently execute them (they're independent)